pkg/util/provider/machinecontroller/machine.go

```diff
@@ -585,6 +600,55 @@ func (c *controller) triggerCreationFlow(ctx context.Context, createMachineReque
 	return machineutils.LongRetry, nil
 }
 
+func (c *controller) initializeMachine(ctx context.Context, machine *v1alpha1.Machine, machineClass *v1alpha1.MachineClass, secret *corev1.Secret) (machineutils.RetryPeriod, error) {
+	req := &driver.InitializeMachineRequest{
+		Machine:      machine,
+		MachineClass: machineClass,
+		Secret:       secret,
+	}
+	klog.V(3).Infof("Initializing VM instance for Machine %q", machine.Name)
+	resp, err := c.driver.InitializeMachine(ctx, req)
+	if err != nil {
+		errStatus, ok := status.FromError(err)
+		if !ok {
+			klog.Errorf("Error occurred while decoding machine error for machine %q: %s", machine.Name, err)
+			return machineutils.MediumRetry, err
+		}
+		klog.Errorf("Error occurred while initializing VM instance for machine %q: %s", machine.Name, err)
+		switch errStatus.Code() {
+		case codes.Unimplemented:
+			klog.V(3).Infof("Provider does not support Driver.InitializeMachine - skipping VM instance initialization for %q.", machine.Name)
+			return 0, nil
+		case codes.NotFound:
+			klog.Warningf("No VM instance found for machine %q. Skipping VM instance initialization.", machine.Name)
+			return 0, nil
+		case codes.Canceled, codes.Aborted:
+			klog.Warningf("VM instance initialization for machine %q was aborted/cancelled.", machine.Name)
+			return 0, nil
+		}
+		retryPeriod, _ := c.machineStatusUpdate(
+			ctx,
+			machine,
+			v1alpha1.LastOperation{
+				Description:    fmt.Sprintf("Provider error: %s. %s", err.Error(), machineutils.InstanceInitialization),
+				ErrorCode:      errStatus.Code().String(),
+				State:          v1alpha1.MachineStateFailed,
```

Comment:
The `triggerCreationFlow` is retried with short retry. when `machineStatusUpdate` updates operation state to `Failed`. Please don't confuse `Phase` with `State`. The state represents the _operation state_.  and not the machine phase.

I will add `codes.InvalidArgument` to the ignore cases. so that nothing is attempted further in `InitializeMachine`.  Though is this right ? I thought it best to retry here in case there was some issue reading machine class or provider spec which could get fixed by next iteration.

For `codes.Internal` the retry is indeed attempted.  I can add some comments to make this logic clearer. I wish we had better utility methods.

Reply:
Ok now have separate `updateRetryPeriod, updateErr`